pallets/liquidity-mining/src/tests/invariants.rs

```diff
@@ -886,3 +900,404 @@ fn invariant_4() {
+
+		runner
+			.run(
+				&(
+					arb_deposit2(),
+					blocks_offset_range,
+					deposit_idx_range,
+					planned_yielding_periods(),
+					percentage(),
+				),
+				|(d, blocks_offset, deposit_idx, planned_yielding_period, percent)| {
+					deposits.borrow_mut().push(d.clone());
+
+					let yield_per_period = Perquintill::from_percent(percent);
+
+					//Act
+					let _ = with_transaction(|| {
+						LiquidityMining::update_global_farm(
+							d.global_farm_id,
+							planned_yielding_period,
+							yield_per_period,
+							1_000,
+						)
+						.unwrap();
+						TransactionOutcome::Commit(DispatchResult::Ok(()))
+					});
+					let g_farm = LiquidityMining::global_farm(d.global_farm_id).unwrap();
+					distributed_before_update.borrow_mut()[d.global_farm_id as usize] =
+						g_farm.accumulated_paid_rewards + g_farm.pending_rewards;
+
+					let _ = with_transaction(|| {
+						assert_ok!(LiquidityMining::deposit_lp_shares(
+							d.global_farm_id,
+							d.yield_farm_id,
+							d.amm_pool_id,
+							d.shares,
+							|_, _, _| -> Result<Balance, DispatchError> { Ok(d.valued_shares) }
+						));
+
+						set_block_number(mock::System::block_number() + blocks_offset);
+
+						//claim rewards only if deposit exists
+						if deposit_idx < deposits.borrow().len() {
+							let d = &deposits.borrow()[deposit_idx];
+							let deposit_id = deposit_idx as u128 + 1;
+
+							assert_ok!(LiquidityMining::claim_rewards(ALICE, deposit_id, d.yield_farm_id, true));
+						}
+
+						TransactionOutcome::Commit(DispatchResult::Ok(()))
+					});
+
+					//Assert:
+					G_FARMS.with(|v| {
+						v.borrow().clone().into_iter().for_each(|gf| {
+							let g_farm_balance = Tokens::free_balance(
+								gf.reward_currency,
+								&LiquidityMining::farm_account_id(gf.id).unwrap(),
+							);
+							let g_farm_1 = LiquidityMining::global_farm(gf.id).unwrap();
+
+							//1.1 assert
+							let s_1 = g_farm_balance + g_farm_1.accumulated_paid_rewards + g_farm_1.pending_rewards;
+							//NOTE: This should be precise.
+							assert_eq!(gf.total_rewards, s_1);
+
+							//1.2 assert
+							let s_1: u128 = g_farm_1.max_reward_per_period * g_farm_1.planned_yielding_periods as u128
+								+ distributed_before_update.borrow()[gf.id as usize];
+
+							//NOTE: Approax becasue of div in max_reward_per_period calculation.
+							assert!(
+								gf.total_rewards >= s_1,
+								"total_rewards >= distributed + max_reward_per_period * planned_yielding_periods"
+							);
+							assert_eq_approx!(
+								gf.total_rewards,
+								s_1,
+								10_000_000,
+								"total_rewards >= distributed + max_reward_per_period * planned_yielding_periods"
+							);
+						})
+					});
+
+					Ok(())
+				},
+			)
+			.unwrap();
+	});
+}
+
+#[test]
+//Update global farm for invariants 1,2,3
+//https://www.notion.so/Liquidity-mining-spec-b30ccfe470a74173b82c3702b1e8fca1#422dea2e23744859baeb704dbdb3caca
+//
+fn update_global_farm_invariant_1_2_3() {
```

Comment:
invariant 1 has its' own test so this name is not correct, please fix it also in comment and `test_name`